---
title: "16S reanalysis"
author: "Hadrien Gourl√©"
date: "4/2/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This Rmarkdown document is part of a reanalysis effort of 13 metabarcoding studies in coral reefs.

The data have been reanalysed with [dada2](https://benjjneb.github.io/dada2/index.html) and this document aims a visualising and interpreting the results.

The code to run the whole pipeline can be found [here](https://github.com/HadrienG/2018_coral_16S)

## Load and tidy the data

```{r load_packages}
library(tidyverse)
library(ggsci)
library(RColorBrewer)
library(gridExtra)
library(reticulate)
```

```{r color_palettes}
# colors_13 <-  c("#ffa755", "#3e0a7c", "#591300", "#006cbe", "#80002e",
#                 "#02e0be", "#944400", "#ff70a5", "#00601b", "#d53694",
#                 "#46c35b", "#b00061", "#ffca7f")
colors_13 <- c('#a6cee3', '#1f78b4', '#b2df8a', '#33a02c', '#fb9a99',
               '#e31a1c', '#fdbf6f', '#ff7f00', '#cab2d6', '#6a3d9a',
               '#ffff99','#b15928', '#808080')
# color <-  grDevices::colors()[grep('gr(a|e)y', grDevices::colors(), invert = T)]
colors_26 <-  c("#ffa755", "#3e0a7c", "#591300", "#006cbe", "#80002e",
                "#02e0be", "#944400", "#ff70a5", "#00601b", "#d53694",
                "#46c35b", "#b00061", "#ffca7f", "#00327f", "#daa61e",
                "#cb729a", "#62ed9b", "#9a0081", "#677300", "#b553c4",
                "#006b3b", "#8770e9", "#c51e38", "#00bcfb", "#cc4128",
                "#fa94ff")
colors_25 <- c("#e3e1ff", "#2d1400", "#003f63", "#004d30", "#98ffea",
               "#ffa19e", "#64002b", "#6c4000", "#002671", "#9effb5",
               "#007430", "#01af7a", "#e7ff92", "#ffb75f", "#758600", 
               "#0084f4", "#ff5daa", "#8d0089", "#ff72e3", "#d07dff",
               "#ff1751", "#0131bc", "#01ba23", "#d4f200", "#8500c6")
```



```{r tidy_function}
#' merge together the metadata, OTU table and taxa table from dada2 into
#' a tidier tibble ready for plotting with ggplot2
tidy_data <- function(seq, tax, metadata) {
    transposed_seq <- as.data.frame(t(seq)) %>%
        as_tibble() %>%
        rownames_to_column(var = 'OTU')
    taxonomy <- as.data.frame(tax) %>%
        as_tibble() %>%
        rownames_to_column(var = 'OTU')
    taxa_table <- inner_join(taxonomy, transposed_seq, by = 'OTU') %>%
        gather('run', 'count', starts_with('SRR')) %>%
        gather('rank', 'taxonomy', 2:8) %>%
        group_by(taxonomy, run) %>%
        mutate(total = sum(count)) %>%
        ungroup() %>%
        select(-OTU, -count) %>%
        distinct() %>%
        inner_join(metadata, by = 'run') 
    return(taxa_table)
}
```

```{r load_and_tidy}
metadata <- read_csv("data/samples.txt") 
row.names(metadata) <- metadata$run

full_seqtab <- dir('results', pattern = 'seqtab_', full.names = TRUE) %>%
    set_names() %>%
    map(readRDS)
full_taxa <- dir('results', pattern = 'taxa_', full.names = TRUE) %>%
    set_names() %>%
    map(readRDS)

full_data <- map2(full_seqtab, full_taxa, tidy_data, metadata = metadata)

coral_dataset <- bind_rows(full_data)
```

## Descriptive statistics

### Coral microbiome

- [x] How many NAs?
- [ ] To what is it due? Is there as many in the published papers?

```{r nas, warning = FALSE}
coral_dataset %>% filter(rank == 'Kingdom') %>% {
    ggplot(.) +
        geom_point(aes(project, total, color = taxonomy),
                   position = 'jitter', alpha = 0.8) +
        theme_minimal() +
        scale_color_brewer(palette = 'Set1', na.value = 'grey', name = 'Kingdom') +
        theme(axis.text.x = element_text(angle = 30, hjust = 1)) +
        ggtitle('Distribution of Kingdoms across datasets') +
        xlab('Dataset') +
        ylab('Counts (log scale)') +
        scale_y_log10()
}
ggsave(filename='figures/nas.png', device = 'png', dpi = 600)
```

The taxonomic assignments are dominated by NAs, which is expected for 16S environmental samples.
Most papers did not report the fraction of NAs, so it is difficult to compare.

Only two studies (PRJNA310360 and PRJNA313050) contain archaea. *TODO* investigate the primers used.

One study (PRJNA193567) seems to contain very little bacteria

- [ ] How does the distribution of (i) phyla (ii) families and (iii) genera compare across samples?

```{r phyla}
top_phyla <- coral_dataset %>%
    filter(rank == 'Phylum',
           sample != "seawater") %>%
    drop_na() %>%
    group_by(run) %>%
    top_n(5, total) %>%
    filter(total > 20, taxonomy != 'Dinoflagellata', taxonomy != 'Cnidaria',
           taxonomy != 'Arthropoda') %>%
    arrange(total) %>%
    ungroup()

background <- coral_dataset %>%
    filter(rank == 'Phylum',
           sample != "seawater") %>%
    drop_na() %>%
    anti_join(top_phyla) %>%
    group_by(run) %>%
    mutate(count = sum(total)) %>%
    select(-total, -taxonomy) %>%
    dplyr::rename(total = count) %>%
    ungroup() %>%
    add_column(taxonomy = "Other") %>%
    distinct() %>%
    filter(total > 0)

top_phyla <- top_phyla %>%
    bind_rows(background)

ggplot(top_phyla) +
    geom_boxplot(aes(reorder(taxonomy, total, max), total), alpha = 0.8,
             color = 'grey30') +
    geom_jitter(aes(reorder(taxonomy, total, max), total, color = project),
                alpha = 0.7, width = 0.2, size = 0.7) +
    scale_color_manual(values = colors_13) +
    theme_minimal() +
    coord_flip() +
    ylab('Counts (log scale)') +
    xlab('Most abundant phyla') +
    ggtitle('Distribution of phyla') +
    guides(colour = guide_legend(override.aes = list(alpha = 1, size = 2))) +
    scale_y_log10()
ggsave(filename = 'figures/phyla_box.png', device = 'png', dpi = 600)

ggplot(top_phyla) +
    geom_bar(aes(reorder(project, total, max), total, fill = taxonomy),
             stat = 'identity', position = 'fill', alpha = 0.7) +
    scale_fill_manual(values = colors_25) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 30, hjust = 1)) +
    ggtitle('Proportion of phyla per datasets') +
    xlab('Dataset') +
    ylab('Percentage')
ggsave(filename = 'figures/phyla_bar.png', device = 'png', dpi = 600)
```

The most abundant phylum in all but two datasets is proteobacteria.
Let us zoom in on the proteobacteria

```{r}
# load your correct python virtualenv here
# modules needed: taxadb
use_virtualenv("~/.virtualenvs/workspace-w0RVuX47")
```


```{python}
from taxadb.taxid import TaxID
print("this is so cool!")
```


- [ ] Is there a common ground? i.e. is there a shared baseline across samples/datasets?
- [ ] Is there a pattern to see with the location and species info?
- [ ] How do the bleached samples look compared to the others?

```{r bleaching}
ggplot(top_phyla) +
    geom_boxplot(aes(reorder(taxonomy, total, max), total), alpha = 0.2,
                 color = 'grey30') +
    geom_jitter(aes(reorder(taxonomy, total, max), total, color = project),
            alpha = 0.2, width = 0.2, size = 0.7) +
    scale_color_manual(values = colors_13) +
    theme_minimal() +
    coord_flip() +
    ylab('Counts (log scale)') +
    xlab('Most abundant phyla') +
    ggtitle('Distribution of phyla') +
    scale_y_log10() +
    theme(axis.text.y = element_text(angle = 30, hjust = 1, size = 4.5)) +
    guides(colour = guide_legend(override.aes = list(alpha = 1, size = 2))) +
    facet_wrap(~ sample)
ggsave(filename = 'figures/bleached_boxplot.png', device = 'png', dpi = 600)
```


### Reproducibility

- [ ] How much does the sequencing depth differ across projects?
- [ ] How does the numbers of OTUs found in each study compare with the number of OTUs found in the original papers?
