---
title: "16S reanalysis"
author: "Hadrien Gourl√©"
date: "4/2/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This Rmarkdown document is part of a reanalysis effort of 13 metabarcoding studies in coral reefs.

The data have been reanalysed with [dada2](https://benjjneb.github.io/dada2/index.html) and this document aims a visualising and interpreting the results.

The code to run the whole pipeline can be found [here](https://github.com/HadrienG/2018_coral_16S)

## Load and tidy the data

```{r load_packages}
library(tidyverse)
library(ggsci)
library(RColorBrewer)
```


```{r tidy_function}
#' merge together the metadata, OTU table and taxa table from dada2 into
#' a tidier tibble ready for plotting with ggplot2
tidy_data <- function(seq, tax, metadata) {
    transposed_seq <- as.data.frame(t(seq)) %>%
        as_tibble() %>%
        rownames_to_column(var = 'OTU')
    taxonomy <- as.data.frame(tax) %>%
        as_tibble() %>%
        rownames_to_column(var = 'OTU')
    taxa_table <- inner_join(taxonomy, transposed_seq, by='OTU') %>%
        gather('run', 'count', starts_with('SRR')) %>%
        gather('rank', 'taxonomy', 2:8) %>%
        group_by(taxonomy, run) %>%
        mutate(total = sum(count)) %>%
        ungroup() %>%
        select(-OTU, -count) %>%
        distinct() %>%
        inner_join(metadata, by = 'run') 
    return(taxa_table)
}
```

```{r load_and_tidy}
metadata <- read_csv("data/samples.txt") 
row.names(metadata) <- metadata$run

full_seqtab <- dir('results', pattern = 'seqtab_', full.names = TRUE) %>%
    set_names() %>%
    map(readRDS)
full_taxa <- dir('results', pattern = 'taxa_', full.names = TRUE) %>%
    set_names() %>%
    map(readRDS)

full_data <- map2(full_seqtab, full_taxa, tidy_data, metadata = metadata)

coral_dataset <- bind_rows(full_data)
```

## Descriptive statistics

#### Coral microbiome

- [ ] How many NAs? To what is it due?
- [ ] How does the distribution of (i) phyla (ii) families and (iii) genera compare across samples?
- [ ] Is there a common ground? i.e. is there a shared baseline across samples/datasets?
- [ ] Is there a pattern to see with the location and species info?
- [ ] How do the bleached samples look compared to the others?

#### Reproducibility

- [ ] How much does the sequencing depth differ across projects?
- [ ] How does the numbers of OTUs found in each study compare with the number of OTUs found in the original papers?

```{r}
coral_dataset %>% filter(rank == "Phylum",
                         taxonomy %in% c('Proteobacteria',
                                         'Spirochaetes',
                                         'Bacteroidetes',
                                         'Cyanobacteria',
                                         'Actinobacteria')) %>% {
    ggplot(.) +
    geom_boxplot(aes(taxonomy, total, color=project)) +
    theme_minimal() +
    scale_color_brewer(palette = 'Paired') +
    theme(axis.text.x = element_text(angle = 30, hjust = 1)) +
    ggtitle('Distribution of most abundant phyla') +
    xlab('Phylum') +
    ylab('Count')
}
```

