---
title: "Coral Reef microbiome - test reanalysis"
author: "Hadrien Gourl√©"
date: "3/6/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

First we install the necessary packages

```{r install_packages}
library(dada2)
packageVersion("dada2")
```

```{r load_data}
path <- "data"
reads <- sort(list.files(path, pattern = ".fastq", full.names = TRUE))
sample.names <- sapply(strsplit(basename(reads), ".", fixed = TRUE), `[`, 1)
```

```{r qc}
# plotQualityProfile(reads)
filtered_path <- file.path(path, "filtered")
filtered_reads <- file.path(filtered_path,
                            paste0(sample.names, "_filtered.fastq.gz"))
out <- filterAndTrim(reads, filtered_reads, maxN=0, truncQ=2,
                     rm.phix=TRUE, compress=TRUE, multithread=TRUE,
                     truncLen = 450)
errors <- learnErrors(filtered_reads, multithread=TRUE)
# plotErrors(errors, nominalQ=TRUE)
```

```{r derep}
dereps <- derepFastq(filtered_reads)
names(dereps) <- sample.names
```

```{r inference}
dada <- dada(dereps, err = errors, HOMOPOLYMER_GAP_PENALTY=-1, BAND_SIZE=32,
               multithread=TRUE)
seqtab <- makeSequenceTable(dada)
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus",
                                    multithread=TRUE, verbose=TRUE)
```

```{r taxonomy}
taxa <- assignTaxonomy(seqtab.nochim, "db/silva_nr_v132_train_set.fa.gz", multithread=TRUE)
taxa <- addSpecies(taxa, "db/silva_species_assignment_v132.fa.gz")
```

```{r smth}
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
save(seqtab.nochim, file = "results/seqtab.nochim")
save(taxa, file = "results/taxa")
```
